docs/bricks_design.rst

```diff
@@ -0,0 +1,602 @@
+    >>> activations = [tensor.tanh, tensor.tanh, tensor.nnet.softmax]
+    >>> weight_matrices = []
+    >>> bias_vectors = []
+    >>> for i in range(len(dims) - 1):
+    >>>     weight_matrices.append(theano.shared(numpy.random.uniform(
+    >>>         size=(dims[i], dims[i + 1]))))
+    >>>     bias_vectors.append(theano.shared(numpy.zeros(dims[i + 1])))
+    >>> outputs = inputs
+    >>> for dim, activation, weight_matrix, bias_vector in zip(
+    >>>         dims[1:], activations, weight_matrices, bias_vectors):
+    >>>     outputs = activation(outputs.dot(weight_matrix) + bias_vector)
+    >>> 
+    >>> # D's RNN, D2L2v0.1
+    >>> inputs = tensor.tensor3('x')
+    >>> inputs.tag.test_value = numpy.random.uniform(size=(5, 10, 100))
+    >>> dim = 100
+    >>> weight_matrix = theano.shared(numpy.random.uniform(size=(dim, dim)))
+    >>> bias_vector = theano.shared(numpy.random.uniform(size=dim))
+    >>> outputs, _ = theano.scan(
+    >>>     lambda states, _inputs, _weight_matrix: states.dot(_weight_matrix) + _inputs,
+    >>>     sequences=[inputs], non_sequences=[weight_matrix],
+    >>>     outputs_info=tensor.zeros_like(inputs[0]))
+
+The two code snippets above has formed D's brand-new DL toolkit D2L2
+(D Deep Learning Library). He picks 0.1 as the initial version
+identifies. Hooray, an new project has been born!
+
+Well, why not just copy-paste?!
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Indeed, why not? For every new project D can copy the snippets above,
+change dimensions, activation functions, input variables. In fact
+copy-pasting is a viable way of reusing code. In has a big advantage:
+debugging is the easiest when your model is built by a single chunk of
+Theano code. A crash with ``exception_verbosity=high`` will show you
+exactly at which stage the error occured. Likewise, compilation-time
+crashes induced by test values are also very informative when you work
+with monolith models.
+
+Unfortunately, the ease of debugging is the first and the last in the
+list of copy-pasting advantages. The list of disadvantages, is which by
+no means specific to deep learning and Python, is much longer:
+
+1. Any change in the original snippet has to be duplicated in each of
+   dozens (hundreds?) where it was pasted. This is a lot of repetitive
+   work, which means wasted time and high probability of doing it wrong
+   at some point. 
+
+2. If the pasted snippet was modified and then the original snippet was
+   improved, merging the changes in both becomes a non-trivial
+   procedure. Of course, merging is also unavoidable even when a more
+   principled then copy-pasting approach is used, but in the latter case
+   you are assisted by the version control system.
+
+3. The only way to tell if the snippet versions used in two scripts are
+   identical is to read the code. Carefully.
+
+4. The namespaces of the snippet and the place where it is applied
+   become one, which is a potential source of horrible bugs.
+
+5. ...
```

Comment:
"Other people might fix a bug in one of the copies of the snippet and not be aware that there are many others that should be fixed in the rest of the code" 
I think this is an important one in a open-source collaborative project.